### PR TITLE
[java] LooseCoupling - fix false positive with generics

### PR DESCRIPTION
**Pull request by adangel**
I found this while trying to run PMD 7 on PMD 7 sources (dogfood).

It addresses the following false positive:

*   `[INFO] PMD Failure: net.sourceforge.pmd.properties.GenericMultiValuePropertyDescriptor:26 Rule:LooseCoupling Priority:1 Avoid using implementation types like 'C'; use the interface instead.`
    with generics...
    https://github.com/pmd/pmd/blob/f36b99fbd755fbe2f6f4f8b35a4299a936305e26/pmd-core/src/main/java/net/sourceforge/pmd/properties/GenericMultiValuePropertyDescriptor.java#L26

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 24 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/5c4960f4e948450f9bd5e307783afbed/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 45900 violations,<br />
introduces 25638 new violations, 38 new errors and 0 new configuration errors,<br />
removes 182075 violations, 27 errors and 3 configuration errors.<br />
<a href="https://chunk.io/pmd/5c4960f4e948450f9bd5e307783afbed/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/LooseCouplingRule.java [hunk 1]
```diff
@@ -87,4 +91,13 @@ private static boolean isInOverriddenMethodSignature(JavaNode node) {
...
         return false;
     }
+
+    private boolean isTypeParameter(ASTClassOrInterfaceType node) {
+        Set<String> typeParameters = node.getRoot()
+                .descendants(ASTTypeParameter.class)
+                .toStream()
+                .map(ASTTypeParameter::getName)
+                .collect(Collectors.toSet());
+        return typeParameters.contains(node.getSimpleName());
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/LooseCouplingRule.java:101]:
```suggestion
        return node.getTypeMirror().isTypeVariable();
```

**Review by oowekyala — APPROVED** [review]

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/LooseCouplingRule.java [hunk 2]
```diff
@@ -7,6 +7,8 @@
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/LooseCouplingRule.java:11]:
```suggestion
```

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/LooseCouplingRule.java [hunk 3]
```diff
@@ -21,6 +23,7 @@
 import net.sourceforge.pmd.lang.java.ast.ASTSuperExpression;
 import net.sourceforge.pmd.lang.java.ast.ASTThisExpression;
 import net.sourceforge.pmd.lang.java.ast.ASTTypeExpression;
+import net.sourceforge.pmd.lang.java.ast.ASTTypeParameter;
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/LooseCouplingRule.java:26]:
```suggestion
```

**Pull request merged**